Lib/test/test_sys_settrace.py

```diff
@@ -1933,53 +1935,69 @@ def compare_jump_output(self, expected, received):
                        "Received: " + repr(received))
 
     def run_test(self, func, jumpFrom, jumpTo, expected, error=None,
-                 event='line', decorated=False):
+                 event='line', decorated=False, warning=None):
         tracer = JumpTracer(func, jumpFrom, jumpTo, event, decorated)
         sys.settrace(tracer.trace)
         output = []
-        if error is None:
+
+        if warning is None and error is None:
```

Comment:
I think that it may be a good opportunity to use ExitStack.

```
with contextlib.ExitStack() as stack:
    if warning is not None:
        stack.enter_context(...)
    ...
    func(output)
```